### ISSUE TITLE
Missing warning about undefined variable

### ISSUE THREAD

**Issue by ivan-kleshnin** [eslint/eslint#2501]
**.eslintrc**
empty "rules" (default)

**test.js**
```js
import Foo from "foo";

console.log(zzz); // Warns: "zzz" is not defined  no-undef

let bar = new Foo(bar); // Passes: no warning here if variable names on both sides are equal?!
```

**Comment by gyandeeps** [comment]
What version of eslint are you using?
What parser are you using?
Can you paste the output here?

**Comment by michaelficarra** [comment]
This is not a problem. `bar` is already declared in its initialiser.

**Comment by ivan-kleshnin** [comment]
> What version of eslint are you using? What parser are you using?
```
babel-eslint: "^3.1.1",
eslint: "^0.21.0"
```

> This is not a problem. bar is already declared in its initialiser.

What?! `bar` equals to Foo(undefined), e.g. implicit variable creation occurs. 
Thats exactly the case for this warning.

**Comment by nzakas** [comment]
@michaelficarra is correct. let Declarations are preparsed, so `bar` does exist on the right side even though it's declared on the left. `bar` is in the temporal dead zone at that point, and as such, will throw an error when you try to access it. However, it is actually defined.
